Virtual scrolling for conversations list

ShGKme opened the pull request:
PR is made on top of https://github.com/nextcloud/spreed/pull/10051

### ☑️ Resolves

Performance issues:
1. Too many conversations avatars request: 
   Opening a page with 500 conversations immediately makes 500 requests for conversation avatars
2. Initial loading is slow for long chat lists
3. Scrolling a long conversations list and interacting with the menu could be slow on a long conversations list

### 🖼️ Screenshots

Almost no visual changes, except scrolling behavior.

By default scroll to position in virtual scrolling just "teleport" to a new position instantly:

To the end | To unread
---|---
![instant](https://github.com/nextcloud/spreed/assets/25978914/0c55764d-7f8c-4f07-93b9-6b2aea8078f3) | ![instant-2](https://github.com/nextcloud/spreed/assets/25978914/029331b9-b390-4e0a-8996-272038b9ffb7)

It doesn't look good and it is different from the current Talk behavior. And it makes it less understandable from what direction (up/down) the user have moved.

But we cannot just use smooth scrolling... Smooth scrolling from the beginning to the end means - loading **ALL** avatars in between and not very performant in general.

The idea is "imitating" the smooth scrolling:
1. If the new position is far from the current - first instantly move to a distance of one visual list height from the target
2. Then scroll with smooth just for a small number of elements

Scroll close | Scroll a bit far | Scroll from the end of 500
---|---|---
![scroll-close](https://github.com/nextcloud/spreed/assets/25978914/8f26b7be-faaf-4ca2-8cd3-71c626560e16) | ![scroll-far-1](https://github.com/nextcloud/spreed/assets/25978914/f9829726-8afc-427a-898e-f62f87e0b92f) | ![scroll-far-2](https://github.com/nextcloud/spreed/assets/25978914/46021eba-d49f-4641-ac33-e3c8c7d53f3b)

Throw 500 chats scrolling is not super smooth but smooth a bit :D

### 🚧 Tasks

- [x] Choose and add a virtual scroller
  - Currently, one popular from a Vue core team member is used: https://github.com/Akryum/vue-virtual-scroller
  - But I'd switch to our own scroller in future
- [x] Use a virtual scroller for the conversations list
  - [x] Change scrolling and unread mentions handling with the virtual list
- [x] Use `<img>` with implicit `key` instead of `<NcAvatar>` for conversations
  - Without a `key` on `<img>` there is a glitch effect on scrolling because Vue updates `src` of the images. Even though images are cached after the first load, the glitch effect is still visible.
  - It is still a problem for 1-1

With glitching | Without glitching
---|---
![with-glitch](https://github.com/nextcloud/spreed/assets/25978914/699c1220-b547-42a7-bf1a-38a88fb8d6b1) | ![no-glitch](https://github.com/nextcloud/spreed/assets/25978914/4f987474-7cb7-43e8-bae7-4ad669570444)

### 📝 Notes and drawbacks

#### No virtual scrolling in the search results

A virtual scroller is used ONLY for the main conversations list (incl. filtered conversations) and **NOT** for the search results. The search results are more complicated because it is not a homogeneous list but several lists with different elements in between.

#### Glitching effect with 1-1 conversations avatars

To avoid it we need to stop using `<NcAvatar>` for 1-1 (what is problematic with the user status) or add `key` to `<NcAvatar>` and make it more performant on props updating.

### 👀 Alternative solutions 

If we only want to fix the first problem with avatar fetching, we can just implement lazy loading for conversation avatars.

### 🏁 Checklist

- [x] ⛑️ Tests (unit and/or integration) are included or not required
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [x] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required
- [x] 🔖 Capability is added or not needed

Antreesy commented on src/components/LeftSidebar/LeftSidebar.vue:
```diff
@@ -680,55 +698,39 @@ export default {
-					if (this.unreadNum === 0) {
-						this.firstUnreadPos = x.offsetTop
+			this.$nextTick(() => {
+				this.unreadNum = 0
```
As we don't use an exact amount of unread mentions, maybe, transform it to the boolean value? Or discard it and keep only `firstUnreadIndex` for both tracking and navigating purposes?

Antreesy commented on src/components/LeftSidebar/LeftSidebar.vue:
```diff
@@ -680,55 +698,39 @@ export default {
+			this.$nextTick(() => {
+				this.unreadNum = 0
+				const lastConversationInViewport = this.$refs.scroller.getLastItemInViewportIndex()
+				for (let i = this.filteredConversationsList.length - 1; i > lastConversationInViewport; i--) {
```
```suggestion
				for (let i = lastConversationInViewport; i < this.filteredConversationsList.length - 1; i++) {
```
I'd go with another order, as unread mentions have newer lastActivity, and stay higher in the list

ShGKme replied:
I kept the current algorithm, we scroll to the latest (most down) unread mention.

Actually, first I suddenly implemented it as you proposed :D 

And then there was a problem. If you have a long list of unread mentions, then, even if you scroll to the first one, you still have some unread mentions below the viewport.

So 
1. You click on "Unread mentions"
2. It scrolls to the beginning of the long list of unread mentions
3. But you still see the button
4. You click again and move just one move conversation down
5. And then again, until you move to the last one

Antreesy replied:
Is that a feature, like an inbox, to process all one by one?)

ShGKme replied:
Well, we have a filter already to have all unread now :D

So I think the primary feature of this button is to indicate that you have something unread below.

Antreesy commented on src/components/LeftSidebar/ConversationsListVirtual.vue:
```diff
@@ -0,0 +1,123 @@
+
+import 'vue-virtual-scroller/dist/vue-virtual-scroller.css'
+
+import Conversation from './ConversationsList/Conversation.vue'
```
lint

ShGKme replied:
But Linter on the CI says everything is fine...

Antreesy replied:
It's a warning, not an error

ShGKme replied:
🙈

ShGKme commented on src/components/LeftSidebar/LeftSidebar.vue:
```diff
@@ -680,55 +698,39 @@ export default {
+				return
+			}
+
+			await this.$refs.scroller.scrollToConversation(token)
```
Hmm... I need to check it, thanks

ShGKme replied:
Well, the scrolling doesn't work on the `master` as well, because conversation can be joined before the list is loaded. But it helped me to find a regression (https://github.com/nextcloud/spreed/pull/10051#issuecomment-1683867869)

ShGKme commented:
- Rebased onto refactoring PR
- Fixed according to the review - linting, `unreadNum` variable
- Fixed regression with not showing joined conversation during loading

ShGKme commented:
/backport to stable27

Antreesy commented on src/components/LeftSidebar/LeftSidebar.vue:
```diff
@@ -111,17 +111,18 @@
+			<li ref="container" class="left-sidebar__list">
+				<ul class="scroller h-100">
 					<!-- Conversations List -->
 					<template v-if="!isSearching">
```
UPD: using `v-show` and keeping scroller in DOM helps with keeping the scrol position

Antreesy replied:
Still thinking, that it's an issue: try scroll to the bottom and use search from there

ShGKme commented on src/components/LeftSidebar/ConversationsListVirtual.vue:
```diff
@@ -34,10 +34,10 @@
-
 import Conversation from './ConversationsList/Conversation.vue'
 
+import 'vue-virtual-scroller/dist/vue-virtual-scroller.css'
```
> The node_module folder is not shipped, so they would need copying to another location?

No, they wouldn't. `<style src="...">` is a `.vue` file feature, not native (Native `<style>` has no `src` attr). It is supposed to work exactly the same as `import 'path to .css`.

> I know what we're doing it everywhere in the project, bu isn't it possible (and cleaner) to import external styles through [style tag](https://vuejs.org/api/sfc-spec.html#src-imports)?

It is possible. And sounds logical. But I personally would use `import` here:
- We work with styles as with modules and dependencies. For me, it's more visible to import dependencies with import. Lib's styles and lib's are imported close to each other
- It is generally more common practice import css directly in SFC. I've never seen using `<style src` for dependencies. Only for component's css.
- Not applicable in this case, but sometimes IDEs don't properly process using `src` in `SFC`.

ShGKme commented on src/components/LeftSidebar/ConversationsListVirtual.vue:
```diff
@@ -0,0 +1,131 @@
+
+			// Make a trick with "smooth" scroll
+
+			if (index < firstItemIndex) {
```
Thanks.

Fixed, but by correcting the fist and last item calculation.

ShGKme commented on src/components/LeftSidebar/LeftSidebar.vue:
```diff
@@ -686,55 +695,38 @@ export default {
-					return
-				}
+		async scrollToConversation(token) {
+			await this.$nextTick()
```
> When picking a conversation with search, and it's on the bottom, scroller previous position is already dropped, and we scroll through the whole list to the picked one

I'm not sure I understand what you suggest here. Is it about adding `v-show` mentioned in the next comment?

Antreesy replied:
Yes, then destroy and mount again component, previously scroll position is not keeped anywhere.
Patch is required to trigger `abortSearch()` earlier and from proper place, when `scrollToconversation()` will be triggered when join a conversation, and smoothscroll should jump to nearest place and smooth scroll further (your hack)

ShGKme replied:
I removed `resetSearchFilter` at all, and used a direct click event from the child component.

ShGKme commented:
Rebased onto master with new `cachedConversations` feature.

See commits from `fixup! remove unused code`:
- Fix linting errors and remove unused code
- Fix scrolling behavior
- Fix some issues with abort search and scrolling to newly selected conversations

@Antreesy please, have a look.

ShGKme commented:
Join a conversation from search update:

Before | After
---|---
![before](https://github.com/nextcloud/spreed/assets/25978914/688128e4-fabe-4607-90ab-52e99d7907e6) | ![after](https://github.com/nextcloud/spreed/assets/25978914/1c95559d-a3a6-4f99-b62e-ae4c195be308)

Antreesy reviewed (approved):
Tested, better now
Let's rebase onto master and squash, and think again, if `v-show` is applicable

Antreesy commented:
> I'd not mix this feature with virtual scrolling.

Then let's consider it for the follow-up

nickvergessen commented:
So time for the fixup rebase?

ShGKme commented:
Forced pushed with squashed commits and very minor changes aka typos in comments
